File: mixin-base/src/main/java/net/yakclient/mixin/base/registry/pool/ExternalLibRegistryPool.java
package net.yakclient.mixin.base.registry.pool;

import net.yakclient.mixin.base.internal.loader.PackageTarget;

import java.net.URL;

public class ExternalLibRegistryPool extends RegistryPool<URL> {
    @Override
    public Location pool(URL type) {
        final var key = new ExternalLibLocation(type);
        if (!this.pool.containsKey(key)) this.pool.put(key, new PoolQueue<>());

        this.pool.get(key).add(type, null);

        return key;
    }

    @Override
    public PackageTarget register(Location location) {
        return null;
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/registry/pool/ExternalLibRegistryPool.kt
package net.yakclient.mixin.base.registry.pool

import net.yakclient.mixin.base.internal.loader.ContextPoolManager
import net.yakclient.mixin.base.target.JarTarget
import net.yakclient.mixin.base.target.Target
import java.net.URL

class ExternalLibRegistryPool : RegistryPool<URL>() {
    override fun pool(type: URL): Location {
        val key = ExternalLibLocation(type)
        if (!pool.containsKey(key)) pool[key] = PoolQueue()
        pool[key]!!.add(type)
        return key
    }

    override fun register(location: Location): Target {
        require(location is ExternalLibLocation) { "Given location must be a external library to register with the lib registry pool!" }
        val target = JarTarget(location.url)
        ContextPoolManager.applyTarget(target)

        return target;
    }
}